in_toto/models/provenance.py

```diff
@@ -0,0 +1,220 @@
+#!/usr/bin/env python
+
+# Copyright New York University and the in-toto contributors
+# SPDX-License-Identifier: Apache-2.0
+
+"""
+<Program Name>
+  provenance.py
+
+<Author>
+  Furkan Türkal <furkan.turkal@trendyol.com>
+
+<Started>
+  Aug 3, 2021
+
+<Copyright>
+  See LICENSE for licensing information.
+
+<Purpose>
+  Provides a class for provenance metadata which is information gathered when a
+  step of the supply chain is performed.
+"""
+
+import attr
+import securesystemslib.formats
+from in_toto.models.common import Signable
+from in_toto.verifylib import _get_attribute
+
+FILENAME_FORMAT = "{step_name}.{keyid:.8}.provenance"
+FILENAME_FORMAT_SHORT = "{step_name}.provenance"
+UNFINISHED_FILENAME_FORMAT = ".{step_name}.{keyid:.8}.provenance-unfinished"
+UNFINISHED_FILENAME_FORMAT_GLOB = ".{step_name}.{pattern}.provenance-unfinished"
+
+
+@attr.s(repr=False, init=False)
+class Provenance(Signable):
+  """Evidence for a performed step or inspection of the supply chain.
+
+  SPEC: https://github.com/in-toto/attestation/blob/main/spec/predicates/provenance.md
+
+  Provenance is a claim that some entity (builder) produced one or more
+  software artifacts (Statement's subject) by executing some recipe,
+  using some other artifacts as input (materials).
+
+  Attributes:
+    builder: Identifies the entity that executed the recipe, which is
+        trusted to have correctly performed the operation and populated this
+        provenance.
+
+        {
+          "id": "<URI>"
+        }
+
+    recipe: Identifies the configuration used for the build. When combined with
+        materials, this SHOULD fully describe the build, such that re-running
+        this recipe results in bit-for-bit identical output (if the build is
+        reproducible).
+
+        {
+          "type": "<URI>",
+          "definedInMaterial": /* integer */,
+          "entryPoint": "<STRING>",
+          "arguments": { /* object */ },
+          "environment": { /* object */ }
+        }
+
+    metadata: Other properties of the build.
+
+            {
+              "buildInvocationId": "<STRING>",
+              "buildStartedOn": "<TIMESTAMP>",
+              "buildFinishedOn": "<TIMESTAMP>",
+              "completeness": {
+                "arguments": true/false,
+                "environment": true/false,
+                "materials": true/false
+              },
+              "reproducible": true/false
+            }
+
+    materials: The collection of artifacts that influenced the build
+        including sources, dependencies, build tools, base images,
+        and so on.
+
+            {
+              "uri": "<URI>",
+              "digest": { /* DigestSet */ }
+            }
+
+  """
+  _type = attr.ib()
+  builder = attr.ib()
```

Comment:
Strictly a side effect of the delay reviewing this PR (my bad!). We should either add this and mark it as v0.1 (and include v0.2 separately) or just update it to v0.2 and support from there onwards. Either way is fine but v0.2 is mandatory.